docs/generalized_distribution.md

```diff
@@ -0,0 +1,83 @@
+the `CommandDistributionBehavior`. It needs to pass a key, which will be used for the distribution.
+In general this should be the same as command that is getting distributed as it makes it possible to
+correlate the two.
+
+Once the method has been called we write a `CommandDistribution.STARTED` event. After this we will
+write a `CommandDistribution.DISTRIBUTING` event for each of our other partitions. With this we can
+keep track of what is distributed and to which partitions it is distributed.
+The actual distribution to the other partitions is done using the `InterPartitionCommandSender`.
+This sends an SBE message to the other partitions containing the command that's being distributed.
+
+Every other partition will write the command to their log and process it. This will be the same
+processor that's used on the distributing partition. As a result we must make sure we don't
+distribute te command again! For this we have the `isCommandDistributed` method on
+the `TypedRecord`. We can detect if a command is distributed by checking the key. Encoded within the
+key is the partition id that generated it. If this doesn't match the partition id that is processing
+the command, it has been distributed before.
+
+At the end of processing the processor must acknowledge the distribution by calling
+the `acknowledgeCommand` method on the `CommandDistributionBehavior`. This method will decode the
+original partition id from the key and, similar to the distribution, will use
+the `InterPartitionCommandSender` to send acknowledgement back to the original partition.
+
+Once the original partition receives this message it writes the `CommandDistribution.ACKNOWLEDGE`
+command to the log. When processing this it knows the distribution to a specific partition is
+completed. Once all partitions acknowledged the distribution the `CommandDistribution.FINISHED`
+event is written. This indicates the finalization of the distribution. At this point all the record
+metadata and raw record value are removed from the state.
+
+### The `CommandDistributionRecord`
+
+The diagram shows the flow of commands and events. What it doesn't highlight is the record that
+belongs to these.
+
+The `CommandDistributionRecord` is relatively simple. It contains 4 properties which are required to
+distribute a command successfully.
+
+|   Property    |                                                                                           Description                                                                                            |
+|---------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| Partition id  | The value of the partition id is the id of the partition that has created the record.                                                                                                            |
```

Comment:
❌ This is not entirely correct. The value contains the id of the partition that is distributing the command on the `STARTED` and `FINISHED` events. On the other intents (`DISTRIBUTING`, `ACKNOWLEDGE`, and `ACKNOWLEDGED`), the value contains the id of the partition that the command is distributed to.